File: client-ex/lib/slimes_client/decide.ex
defmodule SlimesClient.Decide do
  @moduledoc """
  Estratégia de referência. observação -> ação, pura.

  Prioridade: ataca inimigo desfortificado adjacente; fortifica fronteira
  sob ameaça; expande na direção do inimigo mais próximo; senão passa.

  É O ÚNICO MÓDULO QUE VOCÊ PRECISA EDITAR para jogar.
  """

  import SlimesClient.Observation

  def decide(obs, my_id) do
    enemies = attackable(obs.cells, my_id)
    weak = Enum.filter(enemies, &(&1.fortified == 0))
    border = Enum.filter(border_cells(obs.cells, my_id), &(&1.fortified == 0))
    targets = Enum.sort_by(expandable(obs.cells, my_id), &dist(&1, List.first(enemies)))

    cond do
      weak != [] ->
        %{kind: "attack", x: hd(weak).x, y: hd(weak).y}

      border != [] and enemies != [] ->
        %{kind: "fortify", x: hd(border).x, y: hd(border).y}

      targets != [] ->
        %{kind: "expand", x: hd(targets).x, y: hd(targets).y}

      true ->
        %{kind: "pass"}
    end
  end

  defp dist(_a, nil), do: 0
  defp dist(a, b), do: abs(a.x - b.x) + abs(a.y - b.y)
end

File: client-py/decide.py
"""Estratégia de referência. observação -> ação, pura.

Prioridade: ataca inimigo desfortificado adjacente; fortifica fronteira
sob ameaça; expande na direção do inimigo mais próximo; senão passa.

É O ÚNICO ARQUIVO QUE VOCÊ PRECISA EDITAR para jogar.
"""

from observation import attackable, border_cells, expandable


def decide(obs, my_id):
    enemies = attackable(obs["cells"], my_id)
    weak = [c for c in enemies if c["fortified"] == 0]
    if weak:
        return {"kind": "attack", "x": weak[0]["x"], "y": weak[0]["y"]}

    border = [c for c in border_cells(obs["cells"], my_id) if c["fortified"] == 0]
    if border and enemies:
        return {"kind": "fortify", "x": border[0]["x"], "y": border[0]["y"]}

    targets = expandable(obs["cells"], my_id)
    if targets:
        # expande na direção do inimigo mais próximo, se houver; senão qualquer uma
        targets.sort(key=lambda c: _dist(c, enemies[0] if enemies else None))
        return {"kind": "expand", "x": targets[0]["x"], "y": targets[0]["y"]}

    return {"kind": "pass"}


def _dist(a, b):
    if b is None:
        return 0
    return abs(a["x"] - b["x"]) + abs(a["y"] - b["y"])
